ir/pattern/disjunction.rs

```diff
@@ -57,7 +57,11 @@ impl Disjunction {
     }
 
     pub fn optimise_away_unsatisfiable_branches(&mut self, unsatisfiable: Vec<ScopeId>) {
-        self.conjunctions.retain(|v| !unsatisfiable.contains(&v.scope_id()))
+        let unsatisfiable_branch_ids = self.conjunctions.iter().zip(self.branch_ids.iter())
+            .filter_map(|(conj, branch_id)| unsatisfiable.contains(&conj.scope_id()).then_some(*branch_id))
+            .collect::<Vec<_>>();
+        self.branch_ids.retain(|branch_id| !unsatisfiable_branch_ids.contains(branch_id));
+        self.conjunctions.retain(|conj| !unsatisfiable.contains(&conj.scope_id()))
```

Comment:
Yes, the type-system would have caught this if I had used `Vec<(BranchId, Conjunction)>` instead of two separate vecs. But that change was big when I tried it in the morning.

Reply:
BranchIDs are allocated from a u16 that keeps going up. The way it's implemented now, they will indeed be sequential.
When a DisjunctionExecutor executes a branch with BranchID `x` it will do `output_row.provenance |= 1 << x;`

Reply:
Yes. Unsatisfiable branches will remain 0.

Do we return them in the header?
In the HTTP encoding, no. We only return those indices which are 1. So a provenance of `0..1101` will be encoded as `[0,2,3]`
In the GRPC encoding, I intend to send them over as a little endian byte-array, where one would have to do:
`branch_was_satisfied = byte_array[index / 8] & (1 << index % 8)`

Reply:
ohh unsatisfiable. Sorry, I read unsatisfied for some reason.
That's a good question. We should but idk if the pruning optimisation is taken into account. I'll check.

Reply:
I've moved the `extract_query_structure` call to be before we apply transformations, so we get the query-structure as is. Unsatisfiable branches should now be encoded in the response.